[1/n] Migrate activation kernels to libtorch stable ABI

mikaylagawarecki opened the pull request:
## Purpose

**This change requires 2.10+ to build**

First step towards migration of the cuda wheel to the libtorch stable ABI see https://github.com/vllm-project/vllm/issues/26946. **The benefit of migrating to the libtorch stable ABI is that the stable binary will be able to be built once (with torch 2.10 or greater) and can run in all versions of torch >= 2.10.**

1. `activation_kernels.cu` is migrated to use the stable ABI/API
2. Creates `csrc/torch_bindings_stable.cpp` that registers kernels in activation_kernels.cu via `TORCH_STABLE_LIBRARY_FRAGMENT` to the `_C` namespace
3. Creates a new `_C_stable.so` in addition to the existing `_C.so` that has sources `csrc/activation_kernels.cu`
  `csrc/torch_bindings_stable.cpp` 
  
**I am looking for feedback on the following**

1. The PR separates the stable kernels into a new `_C_stable` target, is this ok?
         a. The rationale is that this extension is built with -DTORCH_TARGET_VERSION=... (ensuring only stable APIs are used)
         b. Kernels in this extension are still registered to the _C namespace for backward compat.
2. The PR requires torch version >=2.10 in order to build `_C_stable`, I'm wondering whether that is alright
          a.  It can be built once with any version >= 2.10 and will be able to run in all versions of torch >= 2.10



## Test Plan
`pytest tests/kernels/core/test_activation.py` 

## Test Result


<img width="1257" height="594" alt="Screenshot 2025-12-17 at 5 47 19 PM" src="https://github.com/user-attachments/assets/525acf3c-48f3-438a-8f31-45079ec87f96" />
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

mikaylagawarecki commented:
cc @youkaichao @mgoin @FlamingoPg @simon-mo

Harry-Chen commented:
@mikaylagawarecki 

> Indeed, the final switch of building a libtorch ABI stable wheel would have to wait till all the relevant files are only using stable APIs, but the progress does not need to be all or nothing. Does it make sense to you if I continue this enablement with a stack that migrates files one by one?

Yes this totally makes sense. So our target is to move all c++ compilation units to `VLLM_STABLE_EXT_SRC`. Do you have any evaluation on this, e.g. can we finish this with PyTorch 2.10, or will we be blocked by some APIs that do not exist yet?

> To clarify what I meant here, I meant that the stable library in `C_stable` registers a `TORCH_LIBRARY_FRAGMENT` with the namespace `_C`. This ensures that the function is still callable from python as `torch._C.silu_and_mul`.
> 
> I don't think there are duplicate symbols because I removed the registrations from the respective TORCH_LIBRARY in the `_C` target

I got it now. Thank you for explaining!

> > We could do this after the release of PyTorch 2.10. I think that is about ~1 month later. And during this period we could run build and test on nightly pytorch
> 
> That sounds good! Is there a guide on how to enable this in CI?

Maybe you can refer to <https://docs.vllm.ai/en/latest/contributing/ci/update_pytorch_version/#test-pytorch-release-candidates-rcs>. And CC @youkaichao and @khluu for more ideas here.

ProExpertProg reviewed (commented):
I think instead of having a bunch of `*_stable.*` files, we should just start a new stable directory.

I see the original high-level RFC, but now that we're doing this, could you actually create a tracking issue with an overall plan? That would make it easier to discuss some of these small (but impactful) decisions.

ProExpertProg commented:
I just worry that instead of migrating everything and keeping cpu separate will increase complexity more than the extra work necessary to just do both. But I'll defer to you. 

I think the stable folder would be cleaner, even if just for the headers. Could you try moving the activation files and see if GitHub deals with it well (I feel like it will)?

mikaylagawarecki committed:
Fix libamd that is linked to

Signed-off-by: Mikayla Gawarecki <mikaylagawarecki@gmail.com>

mikaylagawarecki committed:
[1/n] Migrate activation kernels to torch stable ABI

Cherry-picked from temp4 996b0bb2a with the following additional changes:

- Merged M6 Blackwell 256-bit vectorization (PackedTraits, ld256/st256,
  cc_major branching) with stable ABI APIs (DeviceGuard,
  get_current_cuda_stream, mutable_data_ptr/const_data_ptr,
  VLLM_STABLE_DISPATCH_FLOATING_TYPES)
- Added get_device_prop() cached device properties utility to
  csrc/stable/torch_utils.h for stable ABI compatible device queries
- Added explicit cuda_bf16.h/cuda_fp16.h includes for packed type
  intrinsics (hip equivalents on ROCm)
- Replaced c10::BFloat16/c10::Half with torch::headeronly:: equivalents
- ROCm fix: cuda_compat.h -> ../cuda_compat.h in stable/activation_kernels.cu
- ROCm fix: import _C_stable_libtorch in vllm/platforms/rocm.py

Signed-off-by: Mikayla Gawarecki <mikaylagawarecki@gmail.com>

mikaylagawarecki commented on cmake/utils.cmake:
```diff
@@ -541,7 +541,25 @@ function (define_extension_target MOD_NAME)
     target_link_libraries(${MOD_NAME} PRIVATE torch CUDA::cudart CUDA::cuda_driver ${ARG_LIBRARIES})
   else()
-    target_link_libraries(${MOD_NAME} PRIVATE torch ${TORCH_LIBRARIES} ${ARG_LIBRARIES})
+    # Link against PyTorch's bundled libtorch_hip.so (for DeviceGuard registration)
```
On a RoCM machine, I found that I needed these changes to make sure these two were linked to otherwise the stable DeviceGuard would not work correctly (it seemed that 2 separate hipContexts were created by the raw hip calls that vllm did elsewhere and the libtorch shims that called raw hip APIs)

mikaylagawarecki commented on csrc/ops.h:
```diff
@@ -165,8 +165,6 @@ void persistent_masked_m_silu_mul_quant(
     at::Tensor& y_s,           // (E, T, H//group_size) [OUT]
     bool use_ue8m0);
 
-void mul_and_silu(torch::Tensor& out, torch::Tensor& input);
```
I kept the other declarations because otherwise the codex review was upset at me for deleting declarations for ops that are all defined on CPU as well, but this op isn't defined for CPU so it's deleted

Harry-Chen commented:
Superseded by newer PRs.
